### DISCUSSION TITLE
First drop of PicoGK

### DISCUSSION THREAD

**Discussion by leap71** [leap71/PicoGK#3] [Announcements]
This is our initial release. This first release supports MacOS X out of the box. We will release a Windows version in the coming weeks.

<hr /><em>This discussion was created from the release <a href='https://github.com/leap71/PicoGK/releases/tag/v1.0.0'>First drop of PicoGK</a>.</em>

**Comment by CorrieVS** [comment]
Thanks for making this available!  So excited to play with the code.  For me, the first task is compiling for Linux - will let you know how that goes :)

**Reply by LinKayser** [reply]
There are a few conditional statements that assume it's either Mac or Windows in the CMakeLists files. Other than that, there are a few functions in the C# code that deal with home directory etc. They will probably assert or throw an exception. But since it works well on the Mac, it likely will work fine on LINUX. I have no experience with homebrew on LINUX, but the Mac.sh in Install_Dependencies probably just works.

**Reply by CorrieVS** [reply]
I developed a Docker container and some convenience scripts so Linux developers can play, too.  See https://github.com/CorrieVS/PicoGK_Docker

**Reply by LinKayser** [reply]
Very cool — cannot wait to try it out. I am traveling and just have my ARM Mac with me at the moment, so I can't try it out (Homebrew doesn't provide ARM-compiled libraries). But as soon as I am in front of an Intel machine, I will check out your work. We should find a way to all integrate it into the main branch. There are quite a few people, who are interested it using it on LINUX.

**Reply by CorrieVS** [reply]
Great, I'd love to have testers to put it through some paces!  And expand the developer base :)  Since I was able to make the container and build the runtime library for linux, its possible to include an install script and the generated library file in "main" for a local linux install.  I haven't fully tested the local install yet, but I'll happily make a pull request once I do.

As for changes to allow the docker container dev environment, you could minimally just include the linux generated library file in `Runtime/`.  I felt like it was cleaner to use an external repo and include PicoGK as a subrepo, rather than add Dockerfiles to the main project.  Open to suggestions.

**Reply by CorrieVS** [reply]
I wanted to say the local ubuntu install was a success - its nearly there.  All the examples work with one exception: the time-lapse example.  Unfortunately on my local and Docker installs, this one crashes.  I can run it in Windows, however.  Tracking down the root cause is proving a challenge.  Maybe someone will let me know if it looks familiar - I made an issue on my linux development fork https://github.com/CorrieVS/PicoGK/issues/1

**Reply by LinKayser** [reply]
Can you try if you can save one screenshot in one of the other examples? Viewer.RequestScreenshot is used in the timelapse examples, and maybe this causes an issue?

Lin

**Reply by CorrieVS** [reply]
Hi Lin - I was hoping to dig into this further over the holiday, but haven't had the time yet.  I tried running another example using RequestScreenshot, and got the same error.  So, it does seem related to how glfw is reading pixels.  My guess is some error or difference in reading/writing gl buffers between linux and windows.

**Reply by LinKayser** [reply]
I will try to run your Fork in an Intel LINUX Virtual Machine (I learned the hard way, that you cannot use Homebrew on a virtual machine on an ARM Mac, as it only works with Intel).

Will let you know what I find. In the meantime, no Screenshots for you on LINUX :)

**Reply by CorrieVS** [reply]
lol, thanks very much, Lin!  It would be great if you take a look.

I put the linux version of the repo on the linux branch of my fork https://github.com/CorrieVS/PicoGK/tree/linux .
You'll need the installer, which I put on the PicoGKRuntime fork, since that's where the other installers live https://github.com/CorrieVS/PicoGKRuntime .

**Comment by SajjadSPN4** [comment]
Thanks for sharing, cant wait to test it on Windows. When will .dll file be out? I hope sooner though

**Reply by LinKayser** [reply]
You should be able to compile on Windows already, we had several people successfully do so and email us. The first official drop is a few weeks out.

**Reply by startacco** [reply]
I have some troubles with windows... Probably I did some wrong passage or I miss something..

![image2](https://github.com/leap71/PicoGK/assets/148786147/139a95b8-691d-441d-98a6-f3561c462020)
![image1](https://github.com/leap71/PicoGK/assets/148786147/f257b3ce-7ffa-4a08-9d09-8061d7e484cd)

**Reply by CorrieVS** [reply]
I have this same error on my system, but I am trying to build in linux.  I tried building with .NET SDK 6.0 and 7.0.  If you solve it for Windows, I'm interested in what was the cause.  Also, will be an "Issues" section in this repo in the future?

**Reply by LinKayser** [reply]
<img width="175" alt="image" src="https://github.com/leap71/PicoGK/assets/148448288/206357d4-3f92-44ba-bca3-873442f980aa">
.NET 7.0 should have float.Sqrt(), so not sure why you are seeing this.

The second issue — have you tried just building picogk, and not All_Build? Having said this, I see no reason why ALL_BUILD should not work.

**Reply by CorrieVS** [reply]
Ah, its my mistake - I tried building with .NET 7.0 inside a console app I generated in 6.0.  Recreating the console app with .NET 7.0 solved this issue for me.  So, I'm happy to report I can run the example in Linux :)

**Reply by LinKayser** [reply]
Very cool! Congrats!

**Reply by startacco** [reply]
@CorrieVS you were right. Everything must be .NET 7.0. Then I still have some problems becasue I'm not able to load the picogk.1.0.dll even if it is copied in system32 and/or direcltly pointed

**Reply by LinKayser** [reply]
Unfortunately the picogk.1.0.dll, through openvdb, has a number of additional dependencies, which you also need to copy to System32. We will try to minimize those by compiling their static variants. Check out the "install_dependencies" folder and find the following DLL files, copy those to System32 as well, and it should work.

Unfortunately, I do not have a Windows machine with me at the moment.
Lin

![2023-10-31 21 32 48](https://github.com/leap71/PicoGK/assets/148448288/be8e5008-d54f-4ddc-8589-84a6bf7f9149)

**Reply by startacco** [reply]
OK dll where in the folder 
![image](https://github.com/leap71/PicoGK/assets/148786147/8070b831-b169-4076-87df-98fc5ebefec5)
Now it works. Thank you!

**Reply by LinKayser** [reply]
Excellent - glad to hear!
Lin

**Comment by LinKayser** [comment]
> Also, will be an "Issues" section in this repo in the future?

Under consideration.

**Comment by startacco** [comment]
Everything works in windows. I run all the Leap71 shape kernel examples and I can modify paramenters without problems. The log file is in the folder I created on the PC but I canno't find the stl file. I assume it should be in the same folder of the log file but it isn't.

**Reply by LinKayser** [reply]
To export STL files, you need to call SaveToStlFile on a mesh.

You can create a Mesh from a Voxel field like this:

```c#
Voxels vox .... etc, something that creates a voxel field

Mesh msh = new(vox); // Create a mesh from the Voxels
msh.SaveToStlFile("C:/MySTL.stl"); // or whatever path you like

```

I think none of the ShapeKernel examples export STLs, so you have to add this somewhere if you want a file output. 

If you are looking for the source STLs for some of the PicoGK examples, they are in the Examples/TestFiles subdirectory of the PicoGK folder.

Let me know if I understood correctly,

Best,
Lin
